File: pkg/algo/mixture/test/gmm_test.py
from __future__ import annotations

import argparse
import copy
import json
import os
import tempfile
import sys
import unittest
from pathlib import Path
from typing import Any

BASE_DIR = Path(__file__).resolve().parent
MPLCONFIGDIR = Path(tempfile.gettempdir()) / "gonotelm-mplconfig"
MPLCONFIGDIR.mkdir(parents=True, exist_ok=True)
os.environ.setdefault("MPLCONFIGDIR", str(MPLCONFIGDIR))

import matplotlib.pyplot as plt
import numpy as np
from sklearn.metrics import adjusted_rand_score
from sklearn.mixture import GaussianMixture

TESTDATA_DIR = BASE_DIR / "testdata"
CASES_FILE = TESTDATA_DIR / "gmm_cases.json"


def load_json(path: Path) -> dict[str, Any]:
    return json.loads(path.read_text(encoding="utf-8"))


def load_cases() -> list[dict[str, Any]]:
    cases = json.loads(CASES_FILE.read_text(encoding="utf-8"))
    if not isinstance(cases, list) or not cases:
        raise ValueError(f"invalid gmm cases file: {CASES_FILE}")
    return cases


def get_case_by_name(case_name: str) -> dict[str, Any]:
    for case in load_cases():
        if case["name"] == case_name:
            return case
    raise ValueError(f"case not found: {case_name}")


def load_vectors_from_jsonl(path: Path) -> np.ndarray:
    rows: list[list[float]] = []
    with path.open("r", encoding="utf-8") as file:
        for raw_line in file:
            line = raw_line.strip()
            if not line:
                continue
            rows.append(json.loads(line))
    if not rows:
        raise ValueError(f"jsonl dataset is empty: {path}")
    return np.asarray(rows, dtype=np.float64)


def load_dataset_for_case(case: dict[str, Any]) -> tuple[np.ndarray, np.ndarray | None]:
    source_path = TESTDATA_DIR / case["source_file"]
    kind = case["kind"]
    if kind == "payload_json":
        payload = load_json(source_path)
        vectors = np.asarray(payload["vectors"], dtype=np.float64)
        true_labels = payload.get("true_labels")
        labels = np.asarray(true_labels, dtype=np.int64) if true_labels is not None else None
    elif kind == "jsonl":
        vectors = load_vectors_from_jsonl(source_path)
        labels = None
    else:
        raise ValueError(f"unsupported case kind: {kind}")

    feature_limit = int(case.get("feature_limit", 0))
    if feature_limit > 0 and vectors.shape[1] > feature_limit:
        vectors = vectors[:, :feature_limit].copy()
    return vectors, labels


def canonicalize_snapshot(snapshot: dict[str, Any]) -> dict[str, Any]:
    means = np.asarray(snapshot["means"], dtype=np.float64)
    if means.ndim != 2 or means.shape[0] == 0:
        return snapshot

    sort_keys = tuple(means[:, idx] for idx in range(means.shape[1] - 1, -1, -1))
    order = np.lexsort(sort_keys)
    remap = np.empty(order.shape[0], dtype=np.int64)
    remap[order] = np.arange(order.shape[0], dtype=np.int64)

    labels = np.asarray(snapshot["labels"], dtype=np.int64)
    snapshot["labels"] = remap[labels].tolist()
    snapshot["weights"] = np.asarray(snapshot["weights"], dtype=np.float64)[order].tolist()
    snapshot["means"] = means[order].tolist()
    return snapshot


def fit_sklearn_snapshot(case: dict[str, Any]) -> dict[str, Any]:
    vectors, true_labels = load_dataset_for_case(case)
    model = build_sklearn_model(case, int(case["cluster_count"]))
    labels = model.fit_predict(vectors)

    snapshot: dict[str, Any] = {
        "name": case["name"],
        "dataset_file": case["source_file"],
        "covariance_type": case.get("covariance_type", "full"),
        "cluster_count": int(case["cluster_count"]),
        "labels": labels.tolist(),
        "weights": model.weights_.tolist(),
        "means": model.means_.tolist(),
        "log_likelihood": float(model.score(vectors) * vectors.shape[0]),
        "bic": float(model.bic(vectors)),
        "iterations": int(model.n_iter_),
        "converged": bool(model.converged_),
    }
    if true_labels is not None:
        snapshot["ari_vs_true_labels"] = float(adjusted_rand_score(true_labels, labels))
    return canonicalize_snapshot(snapshot)


def build_sklearn_model(case: dict[str, Any], n_components: int) -> GaussianMixture:
    return GaussianMixture(
        n_components=int(n_components),
        covariance_type=case.get("covariance_type", "full"),
        tol=float(case["tolerance"]),
        reg_covar=float(case["regularization"]),
        max_iter=int(case["max_iterations"]),
        n_init=max(int(case.get("n_init", 1)), 1),
        init_params=case.get("init_params", "kmeans"),
        random_state=int(case["seed"]),
    )


def resolve_auto_range(
    n_samples: int,
    min_components: int | None,
    max_components: int | None,
) -> tuple[int, int]:
    min_k = 1 if min_components is None else int(min_components)
    max_k = min(10, n_samples) if max_components is None else int(max_components)
    if min_k <= 0:
        raise ValueError(f"min_components must be positive, got {min_k}")
    if max_k < min_k:
        raise ValueError(f"max_components must be >= min_components, got min={min_k} max={max_k}")
    if max_k > n_samples:
        raise ValueError(f"max_components={max_k} exceeds n_samples={n_samples}")
    return min_k, max_k


def fit_sklearn_snapshot_auto(
    case: dict[str, Any],
    criterion: str = "bic",
    min_components: int | None = None,
    max_components: int | None = None,
) -> dict[str, Any]:
    criterion = criterion.lower().strip()
    if criterion not in {"bic", "aic"}:
        raise ValueError(f"unsupported criterion: {criterion}")

    vectors, true_labels = load_dataset_for_case(case)
    min_k, max_k = resolve_auto_range(vectors.shape[0], min_components, max_components)
    best: dict[str, Any] | None = None
    best_score = float("inf")

    for n_components in range(min_k, max_k + 1):
        model = build_sklearn_model(case, n_components)
        labels = model.fit_predict(vectors)
        bic = float(model.bic(vectors))
        aic = float(model.aic(vectors))
        score = bic if criterion == "bic" else aic
        if score < best_score:
            best_score = score
            best = {
                "model": model,
                "labels": labels,
                "bic": bic,
                "aic": aic,
                "k": int(n_components),
            }

    if best is None:
        raise RuntimeError("auto selection did not produce a valid model")

    model = best["model"]
    labels = best["labels"]
    snapshot: dict[str, Any] = {
        "name": case["name"],
        "dataset_file": case["source_file"],
        "covariance_type": case.get("covariance_type", "full"),
        "cluster_count": int(best["k"]),
        "labels": labels.tolist(),
        "weights": model.weights_.tolist(),
        "means": model.means_.tolist(),
        "log_likelihood": float(model.score(vectors) * vectors.shape[0]),
        "bic": float(best["bic"]),
        "iterations": int(model.n_iter_),
        "converged": bool(model.converged_),
        "selection_criterion": criterion,
        "selection_score": float(best_score),
        "selection_min_components": int(min_k),
        "selection_max_components": int(max_k),
    }
    if true_labels is not None:
        snapshot["ari_vs_true_labels"] = float(adjusted_rand_score(true_labels, labels))
    return canonicalize_snapshot(snapshot)


def plot_panel(
    ax: plt.Axes,
    vectors: np.ndarray,
    labels: np.ndarray,
    means: np.ndarray,
    title: str,
) -> None:
    ax.scatter(
        vectors[:, 0],
        vectors[:, 1],
        c=labels,
        cmap="tab10",
        s=8,
        alpha=0.65,
        linewidths=0.0,
    )
    ax.scatter(
        means[:, 0],
        means[:, 1],
        c=np.arange(means.shape[0]),
        cmap="tab10",
        marker="X",
        s=140,
        edgecolors="black",
        linewidths=1.1,
    )
    ax.set_title(title, fontsize=10)
    ax.set_xlabel("x")
    ax.set_ylabel("y")
    ax.grid(alpha=0.2, linewidth=0.5)


def _to_plot_space(vectors: np.ndarray, means: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
    if vectors.shape[1] >= 2:
        return vectors[:, :2], means[:, :2]
    if vectors.shape[1] == 1:
        vector_2d = np.hstack([vectors, np.zeros((vectors.shape[0], 1), dtype=vectors.dtype)])
        means_2d = np.hstack([means, np.zeros((means.shape[0], 1), dtype=means.dtype)])
        return vector_2d, means_2d
    raise ValueError("vectors must contain at least one feature")


def render_comparison_plot_from_payload(payload: dict[str, Any]) -> dict[str, Any]:
    case_name = str(payload["case_name"])
    vectors = np.asarray(payload["vectors"], dtype=np.float64)

    sklearn_snapshot = canonicalize_snapshot(payload["sklearn_snapshot"])
    go_snapshot = canonicalize_snapshot(payload["go_snapshot"])

    sklearn_labels = np.asarray(sklearn_snapshot["labels"], dtype=np.int64)
    go_labels = np.asarray(go_snapshot["labels"], dtype=np.int64)
    sklearn_means = np.asarray(sklearn_snapshot["means"], dtype=np.float64)
    go_means = np.asarray(go_snapshot["means"], dtype=np.float64)

    plot_vectors, sklearn_plot_means = _to_plot_space(vectors, sklearn_means)
    _, go_plot_means = _to_plot_space(vectors, go_means)

    ari_go_vs_sklearn = float(adjusted_rand_score(sklearn_labels, go_labels))
    ari_sklearn_vs_true = sklearn_snapshot.get("ari_vs_true_labels")
    ari_go_vs_true = go_snapshot.get("ari_vs_true_labels")

    fig, axes = plt.subplots(1, 2, figsize=(14, 6), dpi=150, sharex=True, sharey=True)

    sklearn_title = (
        "sklearn GaussianMixture\n"
        f"K={int(sklearn_snapshot['cluster_count'])}  "
        f"LL={float(sklearn_snapshot['log_likelihood']):.2f}  "
        f"BIC={float(sklearn_snapshot['bic']):.2f}"
    )
    go_title = (
        "Go GaussianMixture\n"
        f"K={int(go_snapshot['cluster_count'])}  "
        f"LL={float(go_snapshot['log_likelihood']):.2f}  "
        f"BIC={float(go_snapshot['bic']):.2f}"
    )
    if ari_sklearn_vs_true is not None:
        sklearn_title = f"{sklearn_title}  ARI(true)={float(ari_sklearn_vs_true):.4f}"
    if ari_go_vs_true is not None:
        go_title = f"{go_title}  ARI(true)={float(ari_go_vs_true):.4f}"

    plot_panel(
        axes[0],
        plot_vectors,
        sklearn_labels,
        sklearn_plot_means,
        sklearn_title,
    )
    plot_panel(
        axes[1],
        plot_vectors,
        go_labels,
        go_plot_means,
        go_title,
    )
    fig.suptitle(
        f"{case_name}: Go vs sklearn GMM comparison (ARI(go,sk)={ari_go_vs_sklearn:.4f})",
        fontsize=12,
    )
    fig.tight_layout()

    output_file = payload.get("output_file")
    output_path = Path(output_file) if output_file else TESTDATA_DIR / f"gmm_compare_{case_name}.png"
    output_path.parent.mkdir(parents=True, exist_ok=True)
    fig.savefig(output_path, bbox_inches="tight")
    plt.close(fig)

    return {
        "output_file": str(output_path),
        "ari_go_vs_sklearn": ari_go_vs_sklearn,
    }


class TestSklearnGaussianMixtureReference(unittest.TestCase):
    def test_all_cases_are_fittable(self) -> None:
        for case in load_cases():
            with self.subTest(case=case["name"]):
                vectors, _ = load_dataset_for_case(case)
                snapshot = fit_sklearn_snapshot(copy.deepcopy(case))
                self.assertEqual(len(snapshot["labels"]), vectors.shape[0])
                self.assertEqual(snapshot["cluster_count"], int(case["cluster_count"]))
                self.assertTrue(np.isfinite(float(snapshot["log_likelihood"])))
                self.assertTrue(np.isfinite(float(snapshot["bic"])))
                if "ari_vs_true_labels" in snapshot:
                    self.assertGreaterEqual(
                        float(snapshot["ari_vs_true_labels"]),
                        float(case.get("min_ari_to_true", 0.995)),
                    )


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--case", type=str, help="Case name in gmm_cases.json.")
    parser.add_argument(
        "--snapshot-stdout",
        action="store_true",
        help="Output sklearn snapshot JSON to stdout (no intermediate file).",
    )
    parser.add_argument(
        "--snapshot-auto-stdout",
        action="store_true",
        help="Output sklearn auto-selected snapshot JSON to stdout.",
    )
    parser.add_argument(
        "--criterion",
        type=str,
        default="bic",
        help="Selection criterion for --snapshot-auto-stdout: bic or aic.",
    )
    parser.add_argument("--min-components", type=int, help="Auto-selection minimum K.")
    parser.add_argument("--max-components", type=int, help="Auto-selection maximum K.")
    parser.add_argument(
        "--plot-stdin",
        action="store_true",
        help="Read Go/sklearn snapshots from stdin JSON and render comparison plot.",
    )
    args, remaining = parser.parse_known_args()

    if args.snapshot_stdout:
        if not args.case:
            raise ValueError("--snapshot-stdout requires --case")
        case = get_case_by_name(args.case)
        snapshot = fit_sklearn_snapshot(copy.deepcopy(case))
        json.dump(snapshot, sys.stdout, ensure_ascii=False)
        sys.stdout.write("\n")
        return

    if args.snapshot_auto_stdout:
        if not args.case:
            raise ValueError("--snapshot-auto-stdout requires --case")
        case = get_case_by_name(args.case)
        snapshot = fit_sklearn_snapshot_auto(
            copy.deepcopy(case),
            criterion=args.criterion,
            min_components=args.min_components,
            max_components=args.max_components,
        )
        json.dump(snapshot, sys.stdout, ensure_ascii=False)
        sys.stdout.write("\n")
        return

    if args.plot_stdin:
        payload = json.load(sys.stdin)
        result = render_comparison_plot_from_payload(payload)
        json.dump(result, sys.stdout, ensure_ascii=False)
        sys.stdout.write("\n")
        return

    unittest.main(argv=[sys.argv[0], *remaining])


if __name__ == "__main__":
    main()

File: pkg/algo/mixture/test/gmm_test.go
package test

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"math"
	"os"
	"os/exec"
	"path/filepath"
	"slices"
	"strings"
	"testing"
	"time"

	"github.com/gonotelm-lab/gonotelm/pkg/algo/mixture"
	"github.com/gonotelm-lab/gonotelm/pkg/algo/numutil"
)

const (
	casesFileName = "gmm_cases.json"
	uvPythonPath  = "../../.venv/bin/python"
)

type gmmCase struct {
	Name             string  `json:"name"`
	Kind             string  `json:"kind"` // payload_json or jsonl
	SourceFile       string  `json:"source_file"`
	ClusterCount     int     `json:"cluster_count"`
	MaxIterations    int     `json:"max_iterations"`
	Tolerance        float64 `json:"tolerance"`
	Regularization   float64 `json:"regularization"`
	Seed             uint64  `json:"seed"`
	FeatureLimit     int     `json:"feature_limit"`
	MinARIToSklearn  float64 `json:"min_ari_to_sklearn"`
	MinARIToTrue     float64 `json:"min_ari_to_true"`
	LogLikelihoodTol float64 `json:"log_likelihood_tolerance"`
	BICTol           float64 `json:"bic_tolerance"`
	InitParams       string  `json:"init_params"`
	CovarianceType   string  `json:"covariance_type"`
	RequestedNInit   int     `json:"n_init"`
	RequestedPlotDim int     `json:"plot_dim"`
}

type gmmSnapshot struct {
	Name               string      `json:"name"`
	DatasetFile        string      `json:"dataset_file"`
	CovarianceType     string      `json:"covariance_type"`
	ClusterCount       int         `json:"cluster_count"`
	Labels             []int       `json:"labels"`
	Weights            []float64   `json:"weights"`
	Means              [][]float64 `json:"means"`
	LogLikelihood      float64     `json:"log_likelihood"`
	BIC                float64     `json:"bic"`
	Iterations         int         `json:"iterations"`
	Converged          bool        `json:"converged"`
	ARIVsTrueLabels    *float64    `json:"ari_vs_true_labels,omitempty"`
	ARIVsSklearnLabels float64     `json:"ari_vs_sklearn_labels,omitempty"`
}

type payloadDataset struct {
	Name               string      `json:"name"`
	Mode               string      `json:"mode"`
	Vectors            [][]float64 `json:"vectors"`
	TrueLabels         []int       `json:"true_labels"`
	ClusterCount       int         `json:"cluster_count"`
	MaxIterations      int         `json:"max_iterations"`
	Tolerance          float64     `json:"tolerance"`
	Regularization     float64     `json:"regularization"`
	Seed               uint64      `json:"seed"`
	RequestedClusterID int         `json:"requested_cluster_count"`
}

type datasetInput struct {
	Vectors    [][]float64
	TrueLabels []int
}

func TestGaussianMixture_MatchesSklearnBaselines(t *testing.T) {
	testdataDir := "testdata"
	caseFilePath := filepath.Join(testdataDir, casesFileName)
	cases := loadCases(t, caseFilePath)

	for _, tc := range cases {
		tc := tc
		t.Run(tc.Name, func(t *testing.T) {
			data := loadDatasetForCase(t, testdataDir, tc)

			pythonResultCh := make(chan sklearnSnapshotResult, 1)
			go func() {
				snapshot, stderr, err := loadSklearnSnapshotViaUV(tc.Name)
				pythonResultCh <- sklearnSnapshotResult{
					snapshot: snapshot,
					stderr:   stderr,
					err:      err,
				}
			}()

			model, err := mixture.NewGaussianMixture(
				tc.ClusterCount,
				mixture.WithTolerance(tc.Tolerance),
				mixture.WithRegularization(tc.Regularization),
				mixture.WithMaxIterations(tc.MaxIterations),
				mixture.WithNInit(max(tc.RequestedNInit, 1)),
				mixture.WithInitParams(parseInitParams(tc.InitParams)),
				mixture.WithRandomSeed(tc.Seed),
			)
			if err != nil {
				t.Fatalf("new gaussian mixture failed: %v", err)
			}

			evaluation, err := model.FitEvaluate(data.Vectors)
			if err != nil {
				t.Fatalf("fit_evaluate failed: %v", err)
			}

			goSnapshot := gmmSnapshot{
				Name:           tc.Name,
				DatasetFile:    tc.SourceFile,
				CovarianceType: defaultIfEmpty(tc.CovarianceType, "full"),
				ClusterCount:   tc.ClusterCount,
				Labels:         slices.Clone(evaluation.Labels),
				Weights:        evaluation.Weights,
				Means:          evaluation.Means,
				LogLikelihood:  evaluation.LogLikelihood,
				BIC:            evaluation.BIC,
				Iterations:     evaluation.Iterations,
				Converged:      evaluation.Converged,
			}
			if len(data.TrueLabels) > 0 {
				ariTrue := numutil.AdjustedRandIndex(evaluation.Labels, data.TrueLabels)
				goSnapshot.ARIVsTrueLabels = &ariTrue
			}

			pythonResult := <-pythonResultCh
			if pythonResult.err != nil {
				t.Fatalf(
					"run sklearn snapshot via uv failed: %v\nstderr:\n%s",
					pythonResult.err,
					pythonResult.stderr,
				)
			}
			sklearnSnapshot := pythonResult.snapshot
			canonicalizeSnapshot(&sklearnSnapshot)

			canonicalizeSnapshot(&goSnapshot)
			goSnapshot.ARIVsSklearnLabels = numutil.AdjustedRandIndex(goSnapshot.Labels, sklearnSnapshot.Labels)

			if tc.MinARIToSklearn > 0 {
				if goSnapshot.ARIVsSklearnLabels < tc.MinARIToSklearn {
					t.Fatalf(
						"ari(go, sklearn) too low: got %.8f, want >= %.8f",
						goSnapshot.ARIVsSklearnLabels,
						tc.MinARIToSklearn,
					)
				}
			}

			if goSnapshot.ARIVsTrueLabels != nil {
				minARIToTrue := tc.MinARIToTrue
				if minARIToTrue <= 0 {
					minARIToTrue = 0.995
				}
				if *goSnapshot.ARIVsTrueLabels < minARIToTrue {
					t.Fatalf(
						"ari(go, true_labels) too low: got %.8f, want >= %.8f",
						*goSnapshot.ARIVsTrueLabels,
						minARIToTrue,
					)
				}
			}

			logLikelihoodTol := tc.LogLikelihoodTol
			if logLikelihoodTol <= 0 {
				logLikelihoodTol = scaleTolerance(sklearnSnapshot.LogLikelihood, 5e-5, 5e-3)
			}
			bicTol := tc.BICTol
			if bicTol <= 0 {
				bicTol = scaleTolerance(sklearnSnapshot.BIC, 5e-5, 5e-2)
			}

			assertCloseFloat(
				t,
				"log_likelihood",
				goSnapshot.LogLikelihood,
				sklearnSnapshot.LogLikelihood,
				logLikelihoodTol,
			)
			assertCloseFloat(
				t,
				"bic",
				goSnapshot.BIC,
				sklearnSnapshot.BIC,
				bicTol,
			)

			renderComparisonPlotViaUV(
				t,
				gmmComparePlotRequest{
					CaseName:        tc.Name,
					Vectors:         data.Vectors,
					SklearnSnapshot: sklearnSnapshot,
					GoSnapshot:      goSnapshot,
					OutputFile: filepath.Join(
						testdataDir,
						fmt.Sprintf("gmm_compare_%s.png", tc.Name),
					),
				},
			)
		})
	}
}

type sklearnSnapshotResult struct {
	snapshot gmmSnapshot
	stderr   string
	err      error
}

func loadCases(t *testing.T, path string) []gmmCase {
	t.Helper()

	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read cases failed: %v", err)
	}

	var cases []gmmCase
	if err := json.Unmarshal(raw, &cases); err != nil {
		t.Fatalf("unmarshal cases failed: %v", err)
	}
	if len(cases) == 0 {
		t.Fatalf("no gmm test cases found in %s", path)
	}

	for idx, tc := range cases {
		if tc.Name == "" {
			t.Fatalf("case[%d] has empty name", idx)
		}
		if tc.SourceFile == "" {
			t.Fatalf("case[%d] has empty source_file", idx)
		}
		if tc.ClusterCount <= 0 {
			t.Fatalf("case[%d] has invalid cluster_count=%d", idx, tc.ClusterCount)
		}
		if tc.MaxIterations <= 0 {
			t.Fatalf("case[%d] has invalid max_iterations=%d", idx, tc.MaxIterations)
		}
		if tc.Kind != "payload_json" && tc.Kind != "jsonl" {
			t.Fatalf("case[%d] has unsupported kind=%q", idx, tc.Kind)
		}
	}

	return cases
}

func loadDatasetForCase(t *testing.T, testdataDir string, tc gmmCase) datasetInput {
	t.Helper()

	sourcePath := filepath.Join(testdataDir, tc.SourceFile)
	switch tc.Kind {
	case "payload_json":
		payload := loadPayloadDataset(t, sourcePath)
		vectors := truncateFeatureDim(payload.Vectors, tc.FeatureLimit)
		return datasetInput{
			Vectors:    vectors,
			TrueLabels: slices.Clone(payload.TrueLabels),
		}
	case "jsonl":
		vectors := loadJSONLVectors(t, sourcePath)
		vectors = truncateFeatureDim(vectors, tc.FeatureLimit)
		return datasetInput{
			Vectors: vectors,
		}
	default:
		t.Fatalf("unsupported case kind %q", tc.Kind)
		return datasetInput{}
	}
}

func loadPayloadDataset(t *testing.T, path string) payloadDataset {
	t.Helper()

	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read payload dataset failed: %v", err)
	}

	var payload payloadDataset
	if err := json.Unmarshal(raw, &payload); err != nil {
		t.Fatalf("unmarshal payload dataset failed: %v", err)
	}
	if len(payload.Vectors) == 0 {
		t.Fatalf("payload dataset has no vectors: %s", path)
	}
	return payload
}

func loadJSONLVectors(t *testing.T, path string) [][]float64 {
	t.Helper()

	file, err := os.Open(path)
	if err != nil {
		t.Fatalf("open jsonl failed: %v", err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	scanner.Buffer(make([]byte, 64*1024), 8*1024*1024)

	vectors := make([][]float64, 0, 512)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		var vector []float64
		if err := json.Unmarshal([]byte(line), &vector); err != nil {
			t.Fatalf("unmarshal jsonl vector failed: %v", err)
		}
		vectors = append(vectors, vector)
	}
	if err := scanner.Err(); err != nil {
		t.Fatalf("scan jsonl failed: %v", err)
	}
	if len(vectors) == 0 {
		t.Fatalf("jsonl dataset has no vectors: %s", path)
	}
	return vectors
}

func truncateFeatureDim(vectors [][]float64, featureLimit int) [][]float64 {
	return numutil.TruncateFeatureDim(vectors, featureLimit)
}

func loadSklearnSnapshotViaUV(caseName string) (gmmSnapshot, string, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
	defer cancel()

	cmd := exec.CommandContext(
		ctx,
		"uv",
		"run",
		"--python",
		uvPythonPath,
		"python",
		"gmm_test.py",
		"--case",
		caseName,
		"--snapshot-stdout",
	)
	cmd.Dir = "."

	var stdout bytes.Buffer
	var stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	if err := cmd.Run(); err != nil {
		return gmmSnapshot{}, stderr.String(), err
	}

	var snapshot gmmSnapshot
	if err := json.Unmarshal(bytes.TrimSpace(stdout.Bytes()), &snapshot); err != nil {
		return gmmSnapshot{}, stderr.String(), fmt.Errorf("decode sklearn snapshot json failed: %w", err)
	}
	return snapshot, stderr.String(), nil
}

type gmmComparePlotRequest struct {
	CaseName        string      `json:"case_name"`
	Vectors         [][]float64 `json:"vectors"`
	SklearnSnapshot gmmSnapshot `json:"sklearn_snapshot"`
	GoSnapshot      gmmSnapshot `json:"go_snapshot"`
	OutputFile      string      `json:"output_file,omitempty"`
}

func renderComparisonPlotViaUV(t *testing.T, request gmmComparePlotRequest) {
	t.Helper()

	raw, err := json.Marshal(request)
	if err != nil {
		t.Fatalf("marshal plot request failed: %v", err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
	defer cancel()

	cmd := exec.CommandContext(
		ctx,
		"uv",
		"run",
		"--python",
		uvPythonPath,
		"python",
		"gmm_test.py",
		"--plot-stdin",
	)
	cmd.Dir = "."
	cmd.Stdin = bytes.NewReader(raw)

	var stdout bytes.Buffer
	var stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	if err := cmd.Run(); err != nil {
		t.Fatalf("render comparison plot via uv failed: %v\nstderr:\n%s", err, stderr.String())
	}
	if output := strings.TrimSpace(stdout.String()); output != "" {
		t.Logf("python plot output: %s", output)
	}
}

func canonicalizeSnapshot(snapshot *gmmSnapshot) {
	if len(snapshot.Means) == 0 {
		return
	}

	order := make([]int, len(snapshot.Means))
	for idx := range order {
		order[idx] = idx
	}
	slices.SortFunc(order, func(a, b int) int {
		for featureIdx := range snapshot.Means[a] {
			if snapshot.Means[a][featureIdx] < snapshot.Means[b][featureIdx] {
				return -1
			}
			if snapshot.Means[a][featureIdx] > snapshot.Means[b][featureIdx] {
				return 1
			}
		}
		return 0
	})

	remap := make([]int, len(order))
	for newIdx, oldIdx := range order {
		remap[oldIdx] = newIdx
	}

	orderedWeights := make([]float64, len(snapshot.Weights))
	orderedMeans := make([][]float64, len(snapshot.Means))
	for newIdx, oldIdx := range order {
		orderedWeights[newIdx] = snapshot.Weights[oldIdx]
		orderedMeans[newIdx] = slices.Clone(snapshot.Means[oldIdx])
	}

	orderedLabels := make([]int, len(snapshot.Labels))
	for sampleIdx, label := range snapshot.Labels {
		if label < 0 || label >= len(remap) {
			orderedLabels[sampleIdx] = label
			continue
		}
		orderedLabels[sampleIdx] = remap[label]
	}

	snapshot.Weights = orderedWeights
	snapshot.Means = orderedMeans
	snapshot.Labels = orderedLabels
}

func parseInitParams(value string) mixture.InitParams {
	switch value {
	case "", "kmeans":
		return mixture.InitParamsKMeans
	case "random":
		return mixture.InitParamsRandom
	default:
		return mixture.InitParamsKMeans
	}
}

func defaultIfEmpty(value, fallback string) string {
	if value == "" {
		return fallback
	}
	return value
}

func scaleTolerance(reference, relative, absolute float64) float64 {
	return math.Abs(reference)*relative + absolute
}

func assertCloseMatrix(t *testing.T, name string, got, want [][]float64, tolerance float64) {
	t.Helper()

	if len(got) != len(want) {
		t.Fatalf("%s row count mismatch: got=%d want=%d", name, len(got), len(want))
	}
	for rowIdx := range got {
		assertCloseVector(t, fmt.Sprintf("%s[%d]", name, rowIdx), got[rowIdx], want[rowIdx], tolerance)
	}
}

func assertCloseVector(t *testing.T, name string, got, want []float64, tolerance float64) {
	t.Helper()

	if len(got) != len(want) {
		t.Fatalf("%s length mismatch: got=%d want=%d", name, len(got), len(want))
	}
	for idx := range got {
		assertCloseFloat(t, fmt.Sprintf("%s[%d]", name, idx), got[idx], want[idx], tolerance)
	}
}

func assertCloseFloat(t *testing.T, name string, got, want, tolerance float64) {
	t.Helper()

	diff := math.Abs(got - want)
	if diff > tolerance {
		t.Fatalf(
			"%s mismatch: got=%.10f want=%.10f diff=%.10f tol=%.10f",
			name,
			got,
			want,
			diff,
			tolerance,
		)
	}
}
